cs_upload.py

```diff
@@ -120,26 +150,44 @@ def query(
 class CsApiClient:
     graphql_client: GraphQLClient
 
-    def __init__(self, api_key: str, api_url: str, ca_cert: Optional[str]):
-        logger.debug(f"Initializing API client (URL: {api_url}, CA cert: {ca_cert})")
+    def __init__(
+        self,
+        api_key: str,
+        api_url: str,
+        ca_cert: Optional[str],
+        check_ca_cert: bool = True,
+    ):
+        if ca_cert:
+            logger.debug(
+                f"Initializing API client (URL: {api_url}, CA cert: {ca_cert})"
+            )
+        elif not check_ca_cert:
+            logger.debug(f"Initializing API client (URL: {api_url}, no CA check)")
+        else:
+            logger.debug(f"Initializing API client (URL: {api_url})")
         self.graphql_client = GraphQLClient(
-            api_key=api_key, api_url=api_url, ca_cert=ca_cert
+            api_key=api_key,
+            api_url=api_url,
+            ca_cert=ca_cert,
+            check_ca_cert=check_ca_cert,
         )
 
-    def generate_trace_upload_post(self) -> Tuple[str, str]:
+    def generate_trace_upload(self) -> Tuple[str, str, str]:
```

Comment:
**suggestion**: Define a dataclass:

```python

class UploadMethod(enum.Enum):
    PUT = enum.auto()
    POST = enum.auto()

@dataclass(frozen=True)
class UploadInfo:
    object_storage_url: str
    form_data: str
    method: UploadMethod
```

I also suggest parsing the HTTP method here:

- to detect a server issue early: If there's such an error, it's better to abort before any other action is taken by our script.
- to avoid having to check for the right string every time we use this parameter (I found 3 occurrences of the "POST").

Reply:
dataclasses will very slightly reduce our compatibility but that is probably fine (python 3.6 was the minimum before and with dataclasses, it moves to 3.7 IIRC).